src/hotspot/share/nmt/nmtMemoryFileTracker.cpp

```diff
@@ -0,0 +1,168 @@
+/*
+ * Copyright (c) 2024, Oracle and/or its affiliates. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ *
+ */
+
+#include "precompiled.hpp"
+#include "memory/allocation.hpp"
+#include "nmt/nmtCommon.hpp"
+#include "nmt/nmtMemoryFileTracker.hpp"
+#include "nmt/nmtNativeCallStackStorage.hpp"
+#include "nmt/vmatree.hpp"
+#include "runtime/mutex.hpp"
+#include "utilities/growableArray.hpp"
+#include "utilities/nativeCallStack.hpp"
+#include "utilities/ostream.hpp"
+
+MemoryFileTracker* MemoryFileTracker::Instance::_tracker = nullptr;
+PlatformMutex* MemoryFileTracker::Instance::_mutex = nullptr;
+
+MemoryFileTracker::MemoryFileTracker(bool is_detailed_mode)
+: _stack_storage(is_detailed_mode), _devices() {
+}
+
+void MemoryFileTracker::allocate_memory(MemoryFile* device, size_t offset,
+                                            size_t size, MEMFLAGS flag,
+                                            const NativeCallStack& stack) {
+  NativeCallStackStorage::StackIndex sidx = _stack_storage.push(stack);
+  DeviceSpace::Metadata metadata(sidx, flag);
```

Comment:
The goal of using `StackIndex` instead of `NativeCallStack`  is to be able to not have duplicate NCS:s throughout the data structure. That's why we store it in the hashtable instead.